File: server/src/services/agentDeployer.ts
/**
 * Agent Deployer - Deploys the backup agent to remote servers via SSH.
 *
 * Flow:
 * 1. Connect to remote server via SSH with password
 * 2. Upload agent binary via SFTP
 * 3. Write config.toml with server URL and server_id
 * 4. Create systemd service
 * 5. Start the service
 * 6. Wait for agent to connect back via WebSocket
 */

import { Client, SFTPWrapper } from 'ssh2';
import fs from 'fs';
import path from 'path';
import { config } from '../config.js';
import { logger } from '../utils/logger.js';
import { isAgentConnected } from '../websocket/agentRegistry.js';

const AGENT_BINARY_PATH = path.resolve(config.dataDir, '..', '..', 'backup-agent', 'target', 'release', 'backup-agent');
const REMOTE_BINARY_PATH = '/usr/local/bin/backup-agent';
const REMOTE_CONFIG_DIR = '/etc/backup-agent';
const REMOTE_CONFIG_PATH = '/etc/backup-agent/config.toml';
const SYSTEMD_SERVICE_PATH = '/etc/systemd/system/backup-agent.service';

interface DeployOptions {
  hostname: string;
  port: number;
  username: string;
  password: string;
  serverId: string;
  serverPort: number;
}

/**
 * Deploy the backup agent to a remote server.
 */
export async function deployAgent(opts: DeployOptions): Promise<void> {
  logger.info({ hostname: opts.hostname }, 'Starting agent deployment');

  // Check if binary exists
  if (!fs.existsSync(AGENT_BINARY_PATH)) {
    throw new Error(`Agent binary not found at ${AGENT_BINARY_PATH}. Build the agent first with: cd backup-agent && cargo build --release`);
  }

  const conn = await connectSSH(opts);

  try {
    // 1. Upload binary (to /tmp first, then sudo mv to final location)
    logger.info({ hostname: opts.hostname }, 'Uploading agent binary...');
    const tmpBinaryPath = `/tmp/backup-agent-upload-${Date.now()}`;
    await uploadFile(conn, AGENT_BINARY_PATH, tmpBinaryPath);
    await execCommand(conn, opts.password, `sudo mv ${tmpBinaryPath} ${REMOTE_BINARY_PATH}`);
    await execCommand(conn, opts.password, `sudo chmod +x ${REMOTE_BINARY_PATH}`);

    // 2. Detect backup server IP as seen from the remote host
    const detectedIp = await detectSourceIp(conn, opts.password);
    const serverUrl = `http://${detectedIp}:${opts.serverPort}`;
    logger.info({ hostname: opts.hostname, serverUrl }, 'Detected backup server URL from SSH connection');

    // 3. Create config directory and write config
    logger.info({ hostname: opts.hostname }, 'Writing agent config...');
    const configContent = generateConfig({ ...opts, serverUrl });
    await execCommand(conn, opts.password, `sudo mkdir -p ${REMOTE_CONFIG_DIR}`);
    await writeRemoteFile(conn, opts.password, REMOTE_CONFIG_PATH, configContent);

    // 3. Create systemd service
    logger.info({ hostname: opts.hostname }, 'Creating systemd service...');
    const serviceContent = generateSystemdService();
    await writeRemoteFile(conn, opts.password, SYSTEMD_SERVICE_PATH, serviceContent);

    // 4. Stop any existing agent and free port 9990
    logger.info({ hostname: opts.hostname }, 'Stopping existing agent (if any)...');
    await execCommand(conn, opts.password, 'sudo systemctl stop backup-agent || true');
    await execCommand(conn, opts.password, 'sudo fuser -k 9990/tcp || true');
    await new Promise(resolve => setTimeout(resolve, 1000));

    // 5. Start the service
    logger.info({ hostname: opts.hostname }, 'Starting agent service...');
    await execCommand(conn, opts.password, 'sudo systemctl daemon-reload');
    await execCommand(conn, opts.password, 'sudo systemctl enable backup-agent');
    await execCommand(conn, opts.password, 'sudo systemctl restart backup-agent');

    // 6. Verify the service started successfully
    await new Promise(resolve => setTimeout(resolve, 2000));
    const statusResult = await execCommand(conn, opts.password, 'sudo systemctl is-active backup-agent');
    const isActive = statusResult.stdout.trim() === 'active';

    if (!isActive) {
      const journalResult = await execCommand(conn, opts.password, 'sudo journalctl -u backup-agent -n 30 --no-pager');
      const statusDetail = await execCommand(conn, opts.password, 'sudo systemctl status backup-agent --no-pager');
      logger.error(
        { hostname: opts.hostname, status: statusResult.stdout.trim(), journal: journalResult.stdout, statusDetail: statusDetail.stdout },
        'Agent service failed to start'
      );
      throw new Error(`Agent service failed to start (status: ${statusResult.stdout.trim()}). Check server logs for journal output.`);
    }

    logger.info({ hostname: opts.hostname }, 'Agent service is active, waiting for WS connection...');
  } finally {
    conn.end();
  }

  // 5. Wait for agent to connect back via WebSocket
  await waitForAgentConnection(opts.serverId, 30000);

  logger.info({ hostname: opts.hostname }, 'Agent deployed and connected');
}

/**
 * Update the agent on a remote server by sending the update command via WebSocket.
 * The agent will download the new binary from the server and restart.
 */
export function getAgentBinaryPath(): string {
  return AGENT_BINARY_PATH;
}

// ---------------------------------------------------------------------------
// SSH helpers
// ---------------------------------------------------------------------------

function connectSSH(opts: DeployOptions): Promise<Client> {
  return new Promise((resolve, reject) => {
    const conn = new Client();
    conn.on('ready', () => resolve(conn));
    conn.on('error', (err) => reject(err));
    conn.on('keyboard-interactive', (_name, _instructions, _lang, _prompts, finish) => {
      finish([opts.password]);
    });
    conn.connect({
      host: opts.hostname,
      port: opts.port,
      username: opts.username,
      password: opts.password,
      tryKeyboard: true,
      readyTimeout: 30000,
    });
  });
}

function uploadFile(conn: Client, localPath: string, remotePath: string): Promise<void> {
  return new Promise((resolve, reject) => {
    conn.sftp((err, sftp) => {
      if (err) return reject(err);

      const readStream = fs.createReadStream(localPath);
      const writeStream = sftp.createWriteStream(remotePath);

      writeStream.on('close', () => {
        sftp.end();
        resolve();
      });
      writeStream.on('error', (err: Error) => {
        sftp.end();
        reject(err);
      });

      readStream.pipe(writeStream);
    });
  });
}

function execCommand(conn: Client, password: string, cmd: string): Promise<{ code: number; stdout: string; stderr: string }> {
  const sudoCmd = cmd.replace(/\bsudo\b/g, 'sudo -S');

  return new Promise((resolve, reject) => {
    const timeout = setTimeout(() => {
      conn.end();
      reject(new Error(`Command timed out: ${cmd.slice(0, 80)}`));
    }, 30000);

    conn.exec(sudoCmd, (err, stream) => {
      if (err) {
        clearTimeout(timeout);
        return reject(err);
      }

      let stdout = '';
      let stderr = '';

      if (sudoCmd.includes('sudo -S')) {
        stream.write(password + '\n');
      }

      stream.on('data', (data: Buffer) => { stdout += data.toString(); });
      stream.stderr.on('data', (data: Buffer) => {
        const text = data.toString();
        if (text.includes('[sudo] password') || text.includes('Password:')) return;
        stderr += text;
      });
      stream.on('close', (code: number | null) => {
        clearTimeout(timeout);
        resolve({ code: code ?? 1, stdout, stderr });
      });
    });
  });
}

function writeRemoteFile(conn: Client, password: string, remotePath: string, content: string): Promise<void> {
  // Write to /tmp first, then sudo mv to final location
  const tmpPath = `/tmp/backup-agent-deploy-${Date.now()}`;
  return new Promise((resolve, reject) => {
    conn.sftp(async (err, sftp) => {
      if (err) return reject(err);

      const writeStream = sftp.createWriteStream(tmpPath);
      writeStream.on('close', async () => {
        sftp.end();
        try {
          await execCommand(conn, password, `sudo mv ${tmpPath} ${remotePath}`);
          resolve();
        } catch (e) {
          reject(e);
        }
      });
      writeStream.on('error', (err: Error) => {
        sftp.end();
        reject(err);
      });
      writeStream.end(content);
    });
  });
}

async function waitForAgentConnection(serverId: string, timeoutMs: number): Promise<void> {
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    if (isAgentConnected(serverId)) {
      return;
    }
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  logger.warn({ serverId }, 'Agent did not connect within timeout, it may connect later');
}

/**
 * Detect the backup server's IP as seen from the remote host.
 * Uses SSH_CONNECTION env var which contains: client_ip client_port server_ip server_port
 * Falls back to BACKUP_SERVER_IP env var, then to network interface detection.
 */
async function detectSourceIp(conn: Client, password: string): Promise<string> {
  try {
    // SSH_CONNECTION = "client_ip client_port server_ip server_port"
    // The client_ip is the backup server's IP as seen by the remote host
    const result = await execCommand(conn, password, 'echo $SSH_CONNECTION');
    const parts = result.stdout.trim().split(/\s+/);
    if (parts.length >= 1 && parts[0]) {
      const ip = parts[0];
      // Validate it looks like an IP
      if (/^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/.test(ip)) {
        return ip;
      }
    }
  } catch {
    // Fall through to fallback
  }

  // Fallback: use env var or first non-loopback interface
  if (process.env.BACKUP_SERVER_IP) {
    return process.env.BACKUP_SERVER_IP;
  }

  try {
    const os = await import('os');
    const nets = os.networkInterfaces();
    for (const name of Object.keys(nets)) {
      for (const net of nets[name]!) {
        if (net.family === 'IPv4' && !net.internal) {
          return net.address;
        }
      }
    }
  } catch {
    // ignore
  }

  return '127.0.0.1';
}

// ---------------------------------------------------------------------------
// Config / Service generators
// ---------------------------------------------------------------------------

function generateConfig(opts: DeployOptions & { serverUrl: string }): string {
  return `[agent]
id = "${opts.hostname}"
port = 9990
data_dir = "/var/lib/backup-agent"

[server]
url = "${opts.serverUrl}"
token = ""
server_id = "${opts.serverId}"

[sync]
chunk_size = 1048576
compression = "zstd"
compression_level = 3

[log]
level = "info"
output = "stdout"

[daemon]
pid_file = "/var/run/backup-agent.pid"
user = "root"
group = "root"

[performance]
max_concurrent_jobs = 1
io_threads = 4
`;
}

function generateSystemdService(): string {
  return `[Unit]
Description=Backup Agent
After=network.target

[Service]
Type=simple
ExecStart=/usr/local/bin/backup-agent --config /etc/backup-agent/config.toml
Restart=always
RestartSec=5
User=root
Environment=RUST_LOG=info

[Install]
WantedBy=multi-user.target
`;
}

File: backup-server-rs/src/services/agent_deployer.rs
use crate::config::AppConfig;
use crate::ws::agent_registry::AgentRegistry;
use std::io::Read;
use std::path::PathBuf;
use std::sync::Arc;

const REMOTE_BINARY_PATH: &str = "/usr/local/bin/backup-agent";
const REMOTE_CONFIG_DIR: &str = "/etc/backup-agent";
const REMOTE_CONFIG_PATH: &str = "/etc/backup-agent/config.toml";
const SYSTEMD_SERVICE_PATH: &str = "/etc/systemd/system/backup-agent.service";

pub struct DeployOptions {
    pub hostname: String,
    pub port: u16,
    pub username: String,
    pub password: String,
    pub server_id: String,
    pub server_port: u16,
    pub backup_server_ip: Option<String>,
}

pub fn get_agent_binary_path() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../backup-agent/target/release/backup-agent")
}

pub async fn deploy_agent(
    opts: DeployOptions,
    agents: Arc<AgentRegistry>,
) -> anyhow::Result<()> {
    let binary_path = get_agent_binary_path();
    if !binary_path.exists() {
        anyhow::bail!(
            "Agent binary not found at {}. Build with: cd backup-agent && cargo build --release",
            binary_path.display()
        );
    }

    tracing::info!(hostname = %opts.hostname, "Starting agent deployment");

    // Run SSH operations in a blocking task
    let result = tokio::task::spawn_blocking(move || {
        deploy_via_ssh(&opts, &binary_path)
    })
    .await??;

    // Wait for agent to connect
    let server_id = result;
    wait_for_agent_connection(&server_id, agents, 30_000).await;

    tracing::info!("Agent deployed and connected");
    Ok(())
}

fn deploy_via_ssh(opts: &DeployOptions, binary_path: &PathBuf) -> anyhow::Result<String> {
    // Connect via SSH
    let tcp = std::net::TcpStream::connect(format!("{}:{}", opts.hostname, opts.port))?;
    let mut sess = ssh2::Session::new()?;
    sess.set_tcp_stream(tcp);
    sess.handshake()?;

    // Authenticate with password
    sess.userauth_password(&opts.username, &opts.password)
        .map_err(|e| anyhow::anyhow!("SSH authentication failed: {}", e))?;

    if !sess.authenticated() {
        anyhow::bail!("SSH authentication failed");
    }

    // 1. Upload binary via SFTP
    tracing::info!(hostname = %opts.hostname, "Uploading agent binary...");
    let tmp_path = format!("/tmp/backup-agent-upload-{}", std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH).unwrap().as_millis());

    let binary_data = std::fs::read(binary_path)?;
    let sftp = sess.sftp()?;
    let mut remote_file = sftp.create(std::path::Path::new(&tmp_path))?;
    std::io::Write::write_all(&mut remote_file, &binary_data)?;
    drop(remote_file);
    drop(sftp);

    exec_ssh(&sess, &opts.password, &format!("sudo mv {} {}", tmp_path, REMOTE_BINARY_PATH))?;
    exec_ssh(&sess, &opts.password, &format!("sudo chmod +x {}", REMOTE_BINARY_PATH))?;

    // 2. Detect server IP
    let detected_ip = detect_source_ip(&sess, &opts.password, &opts.backup_server_ip)?;
    let server_url = format!("http://{}:{}", detected_ip, opts.server_port);
    tracing::info!(hostname = %opts.hostname, server_url = %server_url, "Detected backup server URL");

    // 3. Write config
    tracing::info!(hostname = %opts.hostname, "Writing agent config...");
    let config_content = generate_config(&opts.hostname, &server_url, &opts.server_id);
    exec_ssh(&sess, &opts.password, &format!("sudo mkdir -p {}", REMOTE_CONFIG_DIR))?;
    write_remote_file(&sess, &opts.password, REMOTE_CONFIG_PATH, &config_content)?;

    // 4. Create systemd service
    tracing::info!(hostname = %opts.hostname, "Creating systemd service...");
    let service_content = generate_systemd_service();
    write_remote_file(&sess, &opts.password, SYSTEMD_SERVICE_PATH, &service_content)?;

    // 5. Stop existing agent
    tracing::info!(hostname = %opts.hostname, "Stopping existing agent...");
    let _ = exec_ssh(&sess, &opts.password, "sudo systemctl stop backup-agent || true");
    let _ = exec_ssh(&sess, &opts.password, "sudo fuser -k 9990/tcp || true");
    std::thread::sleep(std::time::Duration::from_secs(1));

    // 6. Start service
    tracing::info!(hostname = %opts.hostname, "Starting agent service...");
    exec_ssh(&sess, &opts.password, "sudo systemctl daemon-reload")?;
    exec_ssh(&sess, &opts.password, "sudo systemctl enable backup-agent")?;
    exec_ssh(&sess, &opts.password, "sudo systemctl restart backup-agent")?;

    // 7. Verify
    std::thread::sleep(std::time::Duration::from_secs(2));
    let status = exec_ssh(&sess, &opts.password, "sudo systemctl is-active backup-agent")?;
    if status.trim() != "active" {
        let journal = exec_ssh(&sess, &opts.password, "sudo journalctl -u backup-agent -n 30 --no-pager").unwrap_or_default();
        tracing::error!(hostname = %opts.hostname, status = %status.trim(), journal = %journal, "Agent service failed to start");
        anyhow::bail!("Agent service failed to start (status: {})", status.trim());
    }

    tracing::info!(hostname = %opts.hostname, "Agent service is active, waiting for WS connection...");
    Ok(opts.server_id.clone())
}

fn exec_ssh(sess: &ssh2::Session, password: &str, cmd: &str) -> anyhow::Result<String> {
    let sudo_cmd = cmd.replace("sudo", "sudo -S");
    let mut channel = sess.channel_session()?;
    channel.exec(&sudo_cmd)?;

    if sudo_cmd.contains("sudo -S") {
        std::io::Write::write_all(&mut channel, format!("{}\n", password).as_bytes())?;
    }

    let mut stdout = String::new();
    channel.read_to_string(&mut stdout)?;
    channel.wait_close()?;
    Ok(stdout)
}

fn write_remote_file(sess: &ssh2::Session, password: &str, remote_path: &str, content: &str) -> anyhow::Result<()> {
    let tmp_path = format!("/tmp/backup-agent-deploy-{}", std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH).unwrap().as_millis());

    let sftp = sess.sftp()?;
    let mut file = sftp.create(std::path::Path::new(&tmp_path))?;
    std::io::Write::write_all(&mut file, content.as_bytes())?;
    drop(file);
    drop(sftp);

    exec_ssh(sess, password, &format!("sudo mv {} {}", tmp_path, remote_path))?;
    Ok(())
}

fn detect_source_ip(sess: &ssh2::Session, password: &str, fallback_ip: &Option<String>) -> anyhow::Result<String> {
    if let Ok(output) = exec_ssh(sess, password, "echo $SSH_CONNECTION") {
        let parts: Vec<&str> = output.trim().split_whitespace().collect();
        if let Some(ip) = parts.first() {
            let ip = ip.trim();
            if ip.split('.').count() == 4 && ip.split('.').all(|p| p.parse::<u8>().is_ok()) {
                return Ok(ip.to_string());
            }
        }
    }

    if let Some(ip) = fallback_ip {
        return Ok(ip.clone());
    }

    Ok("127.0.0.1".into())
}

async fn wait_for_agent_connection(server_id: &str, agents: Arc<AgentRegistry>, timeout_ms: u64) {
    let start = std::time::Instant::now();
    while start.elapsed().as_millis() < timeout_ms as u128 {
        if agents.is_connected(server_id) {
            return;
        }
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
    }
    tracing::warn!(server_id, "Agent did not connect within timeout, it may connect later");
}

fn generate_config(hostname: &str, server_url: &str, server_id: &str) -> String {
    format!(
        r#"[agent]
id = "{hostname}"
port = 9990
data_dir = "/var/lib/backup-agent"

[server]
url = "{server_url}"
token = ""
server_id = "{server_id}"

[sync]
chunk_size = 1048576
compression = "zstd"
compression_level = 3

[log]
level = "info"
output = "stdout"

[daemon]
pid_file = "/var/run/backup-agent.pid"
user = "root"
group = "root"

[performance]
max_concurrent_jobs = 1
io_threads = 4
"#
    )
}

fn generate_systemd_service() -> String {
    r#"[Unit]
Description=Backup Agent
After=network.target

[Service]
Type=simple
ExecStart=/usr/local/bin/backup-agent --config /etc/backup-agent/config.toml
Restart=always
RestartSec=5
User=root
Environment=RUST_LOG=info

[Install]
WantedBy=multi-user.target
"#
    .to_string()
}
